kernel/boot_info/src/lib.rs

```diff
@@ -0,0 +1,102 @@
+
+#![feature(type_alias_impl_trait)]
+#![no_std]
+
+pub mod multiboot2;
+
+use core::{iter::Iterator, ops::Range};
+use memory_structs::{PhysicalAddress, VirtualAddress};
+
+pub trait MemoryRegion {
+    /// Returns the region's starting physical address.
+    fn start(&self) -> PhysicalAddress;
+
+    /// Returns the region's length.
+    fn len(&self) -> usize;
+
+    /// Returns whether the region can be used by the frame allocator.
+    fn is_usable(&self) -> bool;
+}
+
+pub trait ElfSection {
+    /// Returns the section's name.
+    fn name(&self) -> &str;
+
+    /// Returns the section's starting virtual address.
+    fn start(&self) -> VirtualAddress;
+
+    /// Returns the section's length.
+    fn len(&self) -> usize;
+
+    /// Returns the section's flags.
+    fn flags(&self) -> ElfSectionFlags;
+}
+
+bitflags::bitflags! {
+    /// ELF section flags.
+    pub struct ElfSectionFlags: u64 {
+        /// The section contains data that should be writable during program execution.
+        const WRITABLE = 0x1;
+
+        /// The section occupies memory during the process execution.
+        const ALLOCATED = 0x2;
+
+        /// The section contains executable machine instructions.
+        const EXECUTABLE = 0x4;
+    }
+}
+
+pub trait Module {
+    /// Returns the module's name.
+    fn name(&self) -> Result<&str, &'static str>;
+
+    /// Returns the module's starting physical address.
+    fn start(&self) -> PhysicalAddress;
+
+    /// Returns the module's length.
+    fn len(&self) -> usize;
+}
+
+pub trait BootInformation: 'static {
+    type MemoryRegion<'a>: MemoryRegion;
+    type MemoryRegions<'a>: Iterator<Item = Self::MemoryRegion<'a>>;
+
+    type ElfSection<'a>: ElfSection;
+    type ElfSections<'a>: Iterator<Item = Self::ElfSection<'a>>;
+
+    type Module<'a>: Module;
+    type Modules<'a>: Iterator<Item = Self::Module<'a>>;
+
+    /// Returns the boot information's starting virtual address.
+    fn start(&self) -> Option<VirtualAddress>;
+    /// Returns the boot information's length.
+    fn len(&self) -> usize;
+
+    /// Returns the range of physical addresses at which the kernel code is
+    /// located.
+    fn kernel_memory_range(&self) -> Result<Range<PhysicalAddress>, &'static str>;
+    /// Returns the range of physical addresses at which the bootloader
+    /// information is located.
+    fn bootloader_info_memory_range(&self) -> Result<Range<PhysicalAddress>, &'static str>;
+    /// Returns the range of physical addresses at which the modules are
+    /// located.
+    fn modules_memory_range(&self) -> Result<Range<PhysicalAddress>, &'static str>;
+
+    /// Returns memory regions describing the physical memory.
+    fn memory_regions(&self) -> Result<Self::MemoryRegions<'_>, &'static str>;
+    /// Returns the kernel's ELF sections.
+    fn elf_sections(&self) -> Result<Self::ElfSections<'_>, &'static str>;
+    /// Returns the modules found in the kernel image.
+    fn modules(&self) -> Self::Modules<'_>;
```

Comment:
i haven't tried this, but I _think_ rustc lets you omit the `<'_>` inferred lifetime part to where you can just write `Result<Self::MemoryRegions, ...>`. Nbd either way.

Reply:
Not for associated types.

Reply:
Thanks for clarifying, i figured you knew better 😄